# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

*Source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [29]:
# import pandas, pd is an alias
import pandas as pd
import numpy as np

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
k = pd.Series([2,3,'s','a',12.02, ['srb']])
print(k)
print(s)
print(type(s))

0        2
1        3
2        s
3        a
4    12.02
5    [srb]
dtype: object
0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [7]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [3]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series
type(date_series)


pandas.core.indexes.datetimes.DatetimeIndex

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [15]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
print(s[3])

# accessing elements starting index = 2 till the end
s[2:]

6


2    5
3    6
4    9
dtype: int64

In [34]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

#### Explicitly specifying indices

You might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using `pd.Series()`

In [35]:
# Indexing explicitly
pd.Series([0, 1, 2], index = ['a', 'b', 'c'])
pd.Series([0, 1, 1], index = range(1, 4))

1    0
2    1
3    1
dtype: int64

In [36]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list is equal to the number of elements specified in the series

pd.Series(np.array(range(0,10))**2, index = range(0,10))

0     0
1     1
2     4
3     9
4    16
5    25
6    36
7    49
8    64
9    81
dtype: int32

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [37]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

age    name    occupation
0   22   Vinay      engineer
1   25  Kushal        doctor
2   24    Aman  data analyst
3   28    Saif       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [43]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("global_sales_data/market_fact.csv")

Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [20]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [21]:
market_df.tail()

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [22]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
Ord_id                 8399 non-null object
Prod_id                8399 non-null object
Ship_id                8399 non-null object
Cust_id                8399 non-null object
Sales                  8399 non-null float64
Discount               8399 non-null float64
Order_Quantity         8399 non-null int64
Profit                 8399 non-null float64
Shipping_Cost          8399 non-null float64
Product_Base_Margin    8336 non-null float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.2+ KB


In [40]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [41]:
# Column names
market_df.columns

Index([u'Ord_id', u'Prod_id', u'Ship_id', u'Cust_id', u'Sales', u'Discount',
       u'Order_Quantity', u'Profit', u'Shipping_Cost', u'Product_Base_Margin'],
      dtype='object')

In [25]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [26]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ...,
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]],
      dtype=object)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [44]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [45]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = False)
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [46]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Ord_id  Prod_id   Ship_id    Cust_id       Sales  Discount  \
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798    603.6900      0.00   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798   3872.8700      0.03   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798    243.0500      0.02   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798    127.1600      0.10   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798   2841.4395      0.08   
8393  Ord_5348   Prod_8  SHP_7470  Cust_1798    356.7200      0.07   
8392  Ord_5384   Prod_9  SHP_7519  Cust_1798    181.5000      0.08   
8391  Ord_5435   Prod_4  SHP_7594  Cust_1798   1991.8985      0.07   
8390  Ord_5435  Prod_16  SHP_7594  Cust_1798    316.9900      0.04   
8389  Ord_4444  Prod_13  SHP_6192  Cust_1519    159.4100      0.00   
8388  Ord_4424   Prod_1  SHP_6165  Cust_1519    994.0400      0.03   
8387  Ord_2220   Prod_3  SHP_3019   Cust_851     41.0600      0.04   
8386  Ord_2324   Prod_7  SHP_3189   Cust_851    121.8700      0.07   
8385  Ord_1833   Prod_3  SHP_2527   Cust_637    611.1600      0.04   
8384  Ord_4620   Prod_3  SHP_6435  Cust_1577     59.6200      0.04   
8383  Ord_2722   Prod_1  SHP_3731  Cust_1006   3508.3300      0.04   
8382  Ord_2658   Prod_3  SHP_3636  Cust_1006    120.9800      0.00   
8381  Ord_2696   Prod_4  SHP_3691  Cust_1006   2836.0505      0.01   
8380  Ord_2772   Prod_6  SHP_3807  Cust_1006    308.9200      0.04   
8379  Ord_2722   Prod_5  SHP_3730  Cust_1006   1008.9500      0.04   
8378  Ord_2706   Prod_2  SHP_3705  Cust_1006   1361.9100      0.05   
8377  Ord_2722  Prod_12  SHP_3729  Cust_1006     34.0100      0.00   
8376  Ord_2624   Prod_8  SHP_3590  Cust_1006   1211.0000      0.00   
8375  Ord_2772   Prod_3  SHP_3806  Cust_1006   1413.8200      0.10   
8374  Ord_2658   Prod_5  SHP_3637  Cust_1006   1082.6600      0.08   
8373  Ord_2600  Prod_16  SHP_3560  Cust_1006    106.6400      0.10   
8372  Ord_2772   Prod_9  SHP_3806  Cust_1006     56.9000      0.03   
8371  Ord_2624   Prod_4  SHP_3591  Cust_1006   4924.1350      0.07   
8370  Ord_2696  Prod_13  SHP_3690  Cust_1006     62.7800      0.04   
8369  Ord_3633   Prod_3  SHP_5031  Cust_1274   1169.2600      0.02   
...        ...      ...       ...        ...         ...       ...   
29    Ord_2478  Prod_12  SHP_3395   Cust_931    139.9800      0.07   
28    Ord_2573   Prod_4  SHP_3527   Cust_931   3594.7435      0.05   
27    Ord_2405   Prod_9  SHP_3300   Cust_931   1062.6900      0.01   
26    Ord_2456   Prod_5  SHP_3367   Cust_931    109.5800      0.00   
25    Ord_2335  Prod_13  SHP_3204   Cust_931     38.2600      0.03   
24    Ord_2573   Prod_3  SHP_3525   Cust_931    197.6100      0.08   
23     Ord_996   Prod_7  SHP_1378   Cust_371     35.6400      0.05   
22     Ord_996   Prod_5  SHP_1377   Cust_371   3202.2500      0.09   
21     Ord_996   Prod_6  SHP_1378   Cust_371     81.2500      0.01   
20     Ord_996  Prod_13  SHP_1378   Cust_371     57.1700      0.08   
19     Ord_996  Prod_13  SHP_1378   Cust_371     41.9700      0.05   
18    Ord_4427   Prod_6  SHP_6171  Cust_1521    283.1300      0.08   
17    Ord_4471  Prod_15  SHP_6228  Cust_1521  13255.9300      0.02   
16    Ord_2282   Prod_9  SHP_3122   Cust_839    443.4600      0.06   
15    Ord_2280   Prod_5  SHP_3114   Cust_839    460.6900      0.06   
14    Ord_2207  Prod_10  SHP_3006   Cust_839   1410.9300      0.08   
13    Ord_2207  Prod_11  SHP_3093   Cust_839   3364.2480      0.10   
12    Ord_2978  Prod_16  SHP_4112  Cust_1088    305.0500      0.04   
11    Ord_1925   Prod_6  SHP_2637   Cust_708    465.9000      0.05   
10    Ord_4743   Prod_2  SHP_6615  Cust_1641   4072.0100      0.01   
9     Ord_4725   Prod_6  SHP_6593  Cust_1641     57.2200      0.07   
8     Ord_4725  Prod_13  SHP_6593  Cust_1641    162.0000      0.01   
7     Ord_4725   Prod_4  SHP_6593  Cust_1641   3410.1575      0.10   
6       Ord_31  Prod_12    SHP_41    Cust_26     14.7600      0.01   
5     Ord_5446   Prod_6  SHP_7608  Cust_1818    164.0200      0.03   
4     Ord_5485  Pr

In [21]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                    
Ord_704    Prod_7   SHP_964   Cust_242   2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_704    -1.97           0.70                 0.37  
Ord_149    -3.16           1.49                 0.37  
Ord_4270   -2.73           0.70                 0.81  
Ord_4755   -1.78           0.70                 0.56  
Ord_2252   -2.91           1.49                 0.37

In [22]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [47]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Sales'], ascending = False)

Ord_id Prod_id   Ship_id    Cust_id    Sales  Discount  \
2519  Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04   
1485  Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01   
2506   Ord_262  Prod_9   SHP_358    Cust_66  6553.45      0.03   
1533  Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05   
171   Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09   
7310   Ord_950  Prod_9  SHP_1315   Cust_334  4906.85      0.09   
5797  Ord_5112  Prod_9  SHP_7141  Cust_1729  4273.95      0.05   
6735   Ord_612  Prod_9   SHP_836   Cust_687  3872.38      0.10   
2752  Ord_3443  Prod_9  SHP_4773  Cust_1246  3849.17      0.06   
2913  Ord_3650  Prod_9  SHP_6795  Cust_1683  3353.54      0.07   
5416  Ord_4590  Prod_9  SHP_6385  Cust_1465  3212.97      0.03   
7568  Ord_2774  Prod_9  SHP_3809  Cust_1015  3191.24      0.00   
8239   Ord_793  Prod_9  SHP_1089   Cust_245  2833.19      0.06   
3570  Ord_3071  Prod_9  SHP_4265  Cust_1141  2728.42      0.06   
5592  Ord_1290  Prod_9  SHP_1780   Cust_482  2609.53      0.00   
7074   Ord_555  Prod_9   SHP_754   Cust_202  2568.71      0.10   
1193  Ord_2853  Prod_9  SHP_3927  Cust_1056  2366.18      0.10   
7823  Ord_4370  Prod_9  SHP_6092  Cust_1492  2325.42      0.02   
2772  Ord_3239  Prod_9  SHP_4491  Cust_1205  2300.45      0.02   
6451  Ord_2968  Prod_9  SHP_4092  Cust_1066  2160.83      0.08   
4114  Ord_4989  Prod_9  SHP_6965  Cust_1700  2048.78      0.05   
1102  Ord_4663  Prod_9  SHP_6498  Cust_1612  2026.42      0.03   
901   Ord_1152  Prod_9  SHP_1586   Cust_443  1554.53      0.00   
1600  Ord_5370  Prod_9  SHP_7500  Cust_1809  1527.42      0.04   
2306  Ord_3917  Prod_9  SHP_5439  Cust_1337  1504.01      0.05   
2136  Ord_4595  Prod_9  SHP_6390  Cust_1523  1500.84      0.10   
6587  Ord_3389  Prod_9  SHP_4698  Cust_1236  1484.89      0.03   
4600   Ord_604  Prod_9   SHP_826   Cust_213  1406.64      0.02   
3969  Ord_2737  Prod_9  SHP_3755   Cust_994  1400.91      0.09   
3023  Ord_2461  Prod_9  SHP_3373   Cust_941  1396.75      0.04   
...        ...     ...       ...        ...      ...       ...   
7728  Ord_1666  Prod_1  SHP_2303   Cust_490   101.52      0.02   
5698  Ord_3693  Prod_1  SHP_5118  Cust_1287    96.03      0.10   
7258  Ord_2133  Prod_1  SHP_2910   Cust_808    96.01      0.01   
638   Ord_3579  Prod_1  SHP_4956  Cust_1272    94.99      0.06   
2550  Ord_1654  Prod_1  SHP_2284   Cust_491    90.58      0.06   
8067  Ord_5479  Prod_1  SHP_7655  Cust_1805    88.94      0.05   
3478  Ord_3506  Prod_1  SHP_4857   Cust_395    79.93      0.00   
6337  Ord_1858  Prod_1  SHP_2820   Cust_720    73.86      0.08   
3695    Ord_75  Prod_1    SHP_99    Cust_50    73.07      0.08   
6560  Ord_5107  Prod_1  SHP_7132  Cust_1708    72.72      0.06   
202   Ord_2725  Prod_1  SHP_3736   Cust_997    70.02      0.00   
506   Ord_1933  Prod_1  SHP_2649   Cust_707    69.57      0.05   
2522  Ord_4613  Prod_1  SHP_6420  Cust_1569    68.03      0.02   
5595  Ord_3379  Prod_1  SHP_4685  Cust_1233    66.92      0.01   
4377   Ord_542  Prod_1   SHP_735   Cust_188    63.84      0.02   
1294   Ord_210  Prod_1   SHP_288    Cust_94    63.52      0.08   
3788  Ord_4618  Prod_1  SHP_6432  Cust_1575    58.90      0.03   
3465   Ord_212  Prod_1   SHP_291    Cust_88    58.33      0.02   
3832  Ord_2410  Prod_1  SHP_3307   Cust_963    57.48      0.00   
1123  Ord_4213  Prod_1  SHP_5875  Cust_1428    53.30      0.00   
1861  Ord_4079  Prod_1  SHP_5686  Cust_1386    52.59      0.05   
3416   Ord_362  Prod_1   SHP_481   Cust_144    47.01      0.09   
3069   Ord_130  Prod_1  SHP_3971  Cust_1064    40.06      0.07   
4424  Ord_1971  Prod_1  SHP_3185   Cust_919    32.60      0.03   
3046  Ord_1830  Prod_1  SHP_2524   Cust_634    29.65      0.05   
387   Ord_3713  Prod_1  SHP_5145  Cust_1307    27.83      0.09   
6213  Ord_2746  Prod_1  SHP_3767  Cust_1030    22.61      0.03   
3819   Ord_439  Prod_1   SHP_587   Cust_136    18.73      0.05   
3185  Ord_2314  Prod_1  SHP_3171   Cust_899    1

In [52]:
# Sourabh's workspace:
my_series = pd.Series([i **2 for i in range(1, 900)], index=range(1, 900))
print(my_series)

1           1
2           4
3           9
4          16
5          25
6          36
7          49
8          64
9          81
10        100
11        121
12        144
13        169
14        196
15        225
16        256
17        289
18        324
19        361
20        400
21        441
22        484
23        529
24        576
25        625
26        676
27        729
28        784
29        841
30        900
        ...  
870    756900
871    758641
872    760384
873    762129
874    763876
875    765625
876    767376
877    769129
878    770884
879    772641
880    774400
881    776161
882    777924
883    779689
884    781456
885    783225
886    784996
887    786769
888    788544
889    790321
890    792100
891    793881
892    795664
893    797449
894    799236
895    801025
896    802816
897    804609
898    806404
899    808201
Length: 899, dtype: int64


In [ ]:
)